lib/charms/nginx_ingress_integrator/v0/ingress.py

```diff
@@ -189,40 +241,76 @@ def _on_relation_changed(self, event):
         )
 
         if missing_fields:
-            logger.error(
-                "Missing required data fields for ingress relation: {}".format(
-                    ", ".join(missing_fields)
-                )
+            LOGGER.error(
+                "Missing required data fields for ingress relation: %s",
+                ", ".join(missing_fields),
             )
             self.model.unit.status = BlockedStatus(
-                "Missing fields for ingress: {}".format(", ".join(missing_fields))
+                f"Missing fields for ingress: {', '.join(missing_fields)}"
             )
 
+        # Conform to charm-relation-interfaces.
+        if "name" in ingress_data and "port" in ingress_data:
+            name = ingress_data["name"]
+            port = ingress_data["port"]
+        else:
+            name = ingress_data["service-name"]
+            port = ingress_data["service-port"]
+        event.relation.data[self.model.app]["url"] = f"http://{name}:{port}/"
+
         # Create an event that our charm can use to decide it's okay to
         # configure the ingress.
         self.charm.on.ingress_available.emit()
 
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle a relation-broken event in the ingress relation."""
+        if not self.model.unit.is_leader():
+            return
 
-class IngressProvides(IngressBaseProvides):
-    """This class defines the functionality for the 'provides' side of the 'ingress' relation.
+        # Create an event that our charm can use to remove the ingress resource.
+        self.charm.on.ingress_broken.emit(event.relation)
 
-    Hook events observed:
-        - relation-changed
-    """
+class IngressProxyProvides(Object):
```

Comment:
This isn't in the upstream library I think? We'll need to integrate it there rather than carrying local changes here.

Reply:
Sure! I was thinking to upload it here so everyone can review the modifications first, will file a PR against the ingress-integrator before merging this once everything is OK